Pull request for 2/22 deliverable

caocampb opened the pull request:
Script to read binary seems to work.
Script to write to netCDF seems to work.

It still needs to loop through times and mask the array depending on lat/lon ranges (almost done)
After that, it still needs to be integrated in baselineTiming calling the new method instead of read_data.

Work in progress...

Commit:
Added methods in iomethods.py to read binary MERG and convert to netCDF

Commit:
It now looks through files, but check_for_files is causing an issue for userVariables.filelist

Commit:
read_binary_data now loops through files, but check_for_files is causing an issue for userVariables.filelist

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+        Returns:: None. It writes to a netCDF file with extension .nc
+    '''
+
+    ncdf = netCDF4.Dataset(dirName + '/' + fileName + '.nc', "w", format="NETCDF4")
```
good job finding a convention.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
     print 'Error decoding time string: string does not match a predefined time format'
     return 0
     # **********************************************************************************************************************
+def write_np_array_to_ncdf(lon, lat, t, fileName, dirName):
```
to make the function more generic, consider passing a variable list as parameter. That way, the ch4 variable can be passed in as opposed to hard coded, and other parts of the code (existing parts and new parts) can use this function.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+    ncdf.center = "gsfc"
+
+    time = ncdf.createDimension("time", None)  # Create and set dimensions
+    longitude = ncdf.createDimension("longitude", size=9896)
```
again, to make this more generic, these sizes should be detected according to the dims of lat,lon variables that were passed.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+    timeVariable = ncdf.createVariable("time", "double", dimensions=("time",))
+
+    ch4Variable = ncdf.createVariable("ch4", np.float, dimensions=("time","latitude","longitude",))
+    ch4Variable.comments = "Unknown1 variable comment"
```
Edit, here to passing the list of variables. Maybe a list of dictionaries, where the dict contains all of this info for a variable.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+    ch4Variable.time_statistic = "instantaneous"
+    ch4Variable.missing_value = float(330)
+
+    mergFileName = fileName  # Parse file name to get date and time to set as an attribute of timeVariable
```
the number of times that filename parsing is done here and throughout the code, you should consider making it a function.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+    Read MERG brightness temperature from binary file. Thanks to Brian Wilson for this contribution.
+    File contains two large arrays (2 time epochs: on the hour and the half hour)
+    of temperature (Kelvin) as an unsigned integer byte, offset by 75 so it will fit in the 0-255 range.
+    For documentation, see http://www.cpc.ncep.noaa.gov/products/global_precip/html/README.
```
Include in doc string the assumption that the binary file was unmodified when downloaded. i.e. the dims that are hardcoded in will be this always.

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+
+# **********************************************************************************************************************
+
+def read_binary_data(userVariables):
```
it would be more prudent to simply figure out a way to add the function calls (and any other logic) you created to the current read_data method instead of creating another function for reading binary data (from merg files specifically). I recognize that this was highlighted as a work in progress...nonetheless, remember that the git environment will keep track of changes for you, so rather than creating a new function mirrored from an existing function, fixing it up, then having to merge them into one, it may be prudent to simply start working on the core function. Makes sense?

kwhitehall commented on code/utils.py:
```diff
@@ -615,6 +615,18 @@ def write_c1_grad_script(origsFile):
 
 # *********************************************************************************************************************
 
+def time_in_range(start, end, x):
```
check the function check_times in the class variables.py 
Is there some refactoring that could have been done here to leverage this instead of creating a new function with the same functionality. @cgoodale

cgoodale replied:
@kwhitehall @caocampb In iomethods.get_fileList_for_binaries function where utils.time_in_range is being called we don't have access to an instance of the UserVariables Class where the check_times method could be invoked.

After looking over the iomethods.get_fileList_for_binaries function we could replace this function with the following code:

if (start <= end) and (start <= dateAsDateTime <= end):    #this turns the function above into a one line conditional
    newFileList.append(file)

I haven't fully tested this code, but I think the logic should be sound.  The one benefit to having this function declared on it's own is we could (and should) write tests to double check the logic is complete.

@kwhitehall The one possible refactoring might be to take the check_times method and extract it and put it down in the utils module and have the UserVariables Class invoke it.  It would be a step toward keeping all time checks and validation in the utils module.

kwhitehall commented on code/variables.py:
```diff
@@ -81,7 +81,8 @@ def __init__(self, useJSON):  # useJSON=False): #self):
                          'TRMMdirName': "../datadir/TRMM",
-                         'CEoriDirName': "../datadir/MERG"}
+                         'CEoriDirName': "../datadir/MERG",
+                         'MERGBinaryDirName': "../datadir/MERGBinary"}
```
You shouldn't add another data source here for this. For this use case, the 'CEoriDirName': is "../datadir/MERGBinary"

kwhitehall commented:
@caocampb  and @jominican This is a great effort to kick off! Please address the concerns embedded within. Re: the problem with check_files and filelist variable, it may be related to the TRMM file dates and the merg files you are using. What are the dates of the files being used? The problem may also be related to the new variable name use for the merg binary files. Try using 'CEoriDirName':  "../datadir/MERGBinary" and see what happens.

cgoodale commented on code/iomethods.py:
```diff
@@ -4,16 +4,39 @@
+        dateFromFileName = [token for token in file.split('_') if token.isdigit()]  # Parse date from MERG binary file name
+
+        dateAsDateTime = datetime.strptime(dateFromFileName[0], '%Y%m%d%H')
+        start = datetime.strptime(startTime, '%Y%m%d%H%S')  # Convert start and end times to datetimes to compare with
```
@caocampb The variables start and end need to be moved up out of this loop.  You need to compute them once initially and use them for each loop through the fileList.  The code right now will recompute these values (which shouldn't change) each time a new file is parsed.  This seems like a trivial issue, but it's really important to ensure you aren't doing any extra computation within a for loop.

cgoodale commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+
+    for files in userVariables.filelist:
+        try:
+                bLON, bLAT, t = read_MERG_pixel_file(files)
```
@caocampb Lines 653 - 658 are indented 8 spaces instead of 4.  Try blocks are indented the same amount as any loop or conditional statement.

cgoodale commented on code/iomethods.py:
```diff
@@ -518,5 +544,132 @@ def decode_time_from_string(timeString):
+
+    '''
+
+    global LAT
```
@caocampb @kwhitehall The global LAT and LON vars are not needed because they aren't defined outside of each function.  If they were constants that were defined up at the module level near say the import statements then these are needed.  Currently in the code that isn't the case.

Pylint reported:
W:278, 4: Global variable 'LAT' undefined at the module level (global-variable-undefined)
W:279, 4: Global variable 'LON' undefined at the module level (global-variable-undefined)
W:633, 4: Global variable 'LAT' undefined at the module level (global-variable-undefined)
W:634, 4: Global variable 'LON' undefined at the module level (global-variable-undefined)

So I think we can drop the global declarations of these.

Commit:
Merged read_binary_data into read_data

Commit:
Merged read_binary_data into read_data

Commit:
np_array_to_netcdf4 made more generic, but not quite there yet. Need to figure out a way to assign data from numPy array to variables dynamically. For example, currently it's hardcoded... ch4[:,:,:] = temperatures

Tried to integrate read_binary_data into read_data. Needs testing and a few other things such as adding to the timelist.

kwhitehall commented on code/iomethods.py:
```diff
@@ -267,67 +280,77 @@ def read_data(varName, latName, lonName, userVariables, filelist=None):
 
     # Crash nicely if there are no netCDF files
-    if nfiles == 0:
+    if numberOfFiles == 0:
```
Hey @caocampb, if numberOfFiles is only used here, consider making this a one-liner i.e. if len(userVariables.filelist) == 0:

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +531,99 @@ def decode_time_from_string(timeString):
+def write_np_array_to_ncdf(lon, lat, inputData, fileName, dirName, globalAttrDict, dimensionsDict, variablesDict):
+    '''
+        Purpose:: Convert a numPy array to netCDF
+        Inputs:: lon: A string representing the time units found in the file metadata
```
Please update doc string to indicate what these variables are.

kwhitehall commented on code/iomethods.py:
```diff
@@ -267,67 +280,77 @@ def read_data(varName, latName, lonName, userVariables, filelist=None):
-            timelist.extend(time2store)
-            inputData.extend(tempMaskedValue)
-            thisFile.close()
+        elif fileType == 'binary': # TODO Add logic to add time2store to timelist
```
isn't there already and option for if fileType == 'binary'? Please explain this if-elif statement

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +531,99 @@ def decode_time_from_string(timeString):
     print 'Error decoding time string: string does not match a predefined time format'
     return 0
     # **********************************************************************************************************************
+def write_np_array_to_ncdf(lon, lat, inputData, fileName, dirName, globalAttrDict, dimensionsDict, variablesDict):
```
Interesting approach. I was thinking to pass dictionaries like 
lonDict = {"longitude_type": ("double", ("longitude",), "units": "degrees_east", "long_name": "Longitude", "values": lon}
latDict, timeDict, varDict, etc. @cgoodale Thoughts

Commit:
added documentation

Commit:
Updated Documentation

Commit:
Added slightly more reusable time parsing

Commit:
Added slightly more reusable time parsing

Commit:
Removed some comments and other minor things

Commit:
Removed some comments and other minor things

diegovb committed:
Extracted TRMM reading code, started generalizing it

Commit:
Changed arguments to write_np_array_to_ncdf
Added mask to fulfill lat/lon range in read_data for binary MERG

Commit:
Changed arguments to write_np_array_to_ncdf, needs fixed documnetation

Commit:
Added documentation

Commit:
Removed masking and just used basic indexing to get data within a lat/lon range from numpy array

caocampb committed:
Merge pull request #1 from caocampb/data-read-in

Extracted TRMM reading code, started generalizing it

Commit:
Updated documentation

kwhitehall commented on code/iomethods.py:
```diff
@@ -518,5 +525,164 @@ def decode_time_from_string(timeString):
+
+# **********************************************************************************************************************
+
+if __name__ == '__main__':  # Testing for write_np_array_to_ncdf
```
Please remove this test from the main iomethods.py file. You can consider making it a method in the folder test, maybe create test_iomethods.py and place it there. @cgoodale thoughts?

cgoodale replied:
Hi @kwhitehall and @caocampb, I have to agree with you Kim, but this is way larger than a unit test.  It should still be in the test folder, but we should flag it as a module test or end-to-end test.  I see what Campell was doing by wrapping the code in the if **name** statement because the code won't be run if you import iomethods.py but it will run if you run python iomethods.py.

Campbell my recommendation is to cut this entire code block and paste it into a file called iomethods_module_test.py you will need to import iomethods.py and add in a docstring to explain the purpose of this test, and you should be good to go.  One last thing, in the final version you will want to change that hardcoded path on your machine to something generic like '/path/to/MERG/data/' :+1:

kwhitehall commented on code/iomethods.py:
```diff
@@ -4,40 +4,67 @@
 import string
 import subprocess
 import sys
+import math
```
Why do you need this library?

kwhitehall commented on code/iomethods.py:
```diff
@@ -79,6 +106,8 @@ def check_for_files(dirPath, startTime, endTime, tdelta, tRes):
             elif eachPart in startTime:
                 startTimeInFile += eachPart + '*'
 
+    print startTimeInFile
```
remove this debugging print statement

kwhitehall commented on code/iomethods.py:
```diff
@@ -226,132 +247,123 @@ def create_main_directory(mainDirStr):
+                # Clip the dataset according to user lat, lon coordinates
+                # Mask the data and fill with zeros for later
+                tempRaw = thisFile.variables[varName][:, latminIndex:latmaxIndex, lonminIndex:lonmaxIndex].astype('int16')
+                tempMask = ma.masked_array(tempRaw, mask=(tempRaw > userVariables.T_BB_MAX), fill_value=0)
```
A fill_value of 0 ma not be wise because given the data associated with the data, it may be counted as a valid value. For this particular variable (temperature in Kelvin and most vars in meteorological applications) a fill_value of -999 is better

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
Added tests

Commit:
commit
